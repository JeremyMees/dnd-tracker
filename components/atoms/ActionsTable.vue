<script setup lang="ts">
const props = defineProps<{
  specialAbilities?: (Action | ActionOpen5E)[]
  legendaryActions?: (Action | ActionOpen5E)[]
  actions?: (Action | ActionOpen5E)[]
  reactions?: (Action | ActionOpen5E)[]
}>()

const allActions = computed<Option<Action[]>[]>(() => [
  { label: 'specialAbility', value: props.specialAbilities },
  { label: 'legendaryAction', value: props.legendaryActions },
  { label: 'action', value: props.actions },
  { label: 'reaction', value: props.reactions },
].filter((action): action is Option<Action[]> => !!action.value?.length),
)
</script>

<template>
  <div class="space-y-4">
    <div
      v-for="{ label, value } in allActions"
      :key="label"
      class="space-y-2"
    >
      <p class="head-3">
        {{ $t(`general.${label}`, 2) }}
      </p>
      <Card
        color="secondary"
        class="p-2"
      >
        <ul
          v-for="action in value"
          :key="action.name"
          class="flex w-full flex-col border-b-2 border-secondary py-2 last:border-b-0 last:pb-0 first:pt-0 list-disc"
        >
          <li class="flex flex-wrap gap-x-4 items-center">
            <p class="font-bold">
              {{ action.name }}:
            </p>
            <p class="body-small text-muted-foreground">
              {{ action.desc }}
            </p>
          </li>
          <div
            v-if="action.attack_bonus || action.damage_dice"
            class="flex flex-wrap gap-x-4 items-center mt-2"
          >
            <div
              v-if="action.attack_bonus"
              class="flex flex-wrap gap-x-2 items-center"
            >
              <p class="font-bold">
                To hit:
              </p>
              <p class="body-small text-muted-foreground">
                +{{ action.attack_bonus }}
              </p>
            </div>
            <div
              v-if="action.damage_dice"
              class="flex flex-wrap gap-x-2 items-center"
            >
              <p class="font-bold">
                Dice:
              </p>
              <p class="body-small text-muted-foreground">
                {{ action.damage_dice }}
                <span v-if="action.damage_bonus">
                  +{{ action.damage_bonus }}
                </span>
              </p>
            </div>
          </div>
        </ul>
      </Card>
    </div>
  </div>
</template>
