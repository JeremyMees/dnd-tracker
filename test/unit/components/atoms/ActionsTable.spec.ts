import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'
import ActionsTable from '~/components/atoms/ActionsTable.vue'

interface Props {
  actions?: (Action | ActionOpen5E)[]
  reactions?: (Action | ActionOpen5E)[]
  legendaryActions?: (Action | ActionOpen5E)[]
  specialAbilities?: (Action | ActionOpen5E)[]
}

const props: Props = {
  actions: [
    {
      name: 'Test Action',
      desc: 'Test Description',
      type: 'actions',
      attack_bonus: 10,
      damage_bonus: 10,
      damage_dice: '1d10',
    },
  ],
}

describe('ActionsTable', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(ActionsTable, { props })
    expect(component.html()).toMatchSnapshot()
  })

  it('Should render actions correctly', async () => {
    const component = await mountSuspended(ActionsTable, { props })

    expect(component.html()).toContain('Test Action')
    expect(component.html()).toContain('Test Description')
    expect(component.html()).toContain('To hit')
    expect(component.html()).toContain('+10')
    expect(component.html()).toContain('Dice')
    expect(component.html()).toContain('1d10')
  })
})
