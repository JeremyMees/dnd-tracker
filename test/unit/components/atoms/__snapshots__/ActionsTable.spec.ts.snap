// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionsTable > Should match snapshot 1`] = `
"<div class="space-y-4">
  <div class="space-y-2">
    <p class="head-4">general.action</p>
    <ul class="transition-colors duration-300 border-4 rounded-lg p-4 border-secondary bg-secondary/50 p-2">
      <li class="flex w-full flex-col border-b-2 border-secondary py-2 last:border-b-0 last:pb-0 first:pt-0 list-disc">
        <div class="flex flex-wrap gap-x-4 items-center">
          <p class="font-bold">Test Action: </p>
          <p class="text-sm text-muted-foreground">Test Description</p>
        </div>
        <div class="flex flex-wrap gap-x-4 items-center mt-2">
          <div class="flex flex-wrap gap-x-2 items-center">
            <p class="font-bold"> To hit: </p>
            <p class="text-sm text-muted-foreground"> +10</p>
          </div>
          <div class="flex flex-wrap gap-x-2 items-center">
            <p class="font-bold"> Dice: </p>
            <p class="text-sm text-muted-foreground">1d10 <span> +10</span></p>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>"
`;
